plugin-server/src/utils/db/db.ts

```diff
@@ -997,6 +1002,91 @@ export class DB {
         return kafkaMessages
     }
 
+    public async addPersonOverride(
+        oldPerson: Person,
+        overridePerson: Person,
+        client?: PoolClient
+    ): Promise<ProducerRecord> {
+        const mergedAt = DateTime.now()
+        /** We'll need to do two updates:
+         1. to add an override from oldPerson to override person
+         2. update any entries that have oldPerson as the override person to now also point to the new override person
+
+         TODO: how do we want to be updating oldest_event ? 
+         I'm thinking: we write it if it's not there, but don't update it on conflicts (alternative update to the older date)
+         In the transitive one the same don't update or update to older data (if that's not too complex to do)
+         it's an optimization anyway and squash job might be updating those values anyway and it's just a hint, so ignoring if complex seems better
+         in any case it's important that we use the oldest_event we stored in postgres in CH too.
+        */
+
+        // (1)
+        const {
+            rows: [[version, oldestEvent]],
+        } = await this.postgresQuery(
+            `
+            INSERT INTO posthog_personoverride (
+                team_id, 
+                old_person_id, 
+                override_person_id, 
+                oldest_event,
+                version
+            ) VALUES ($1, $2, $3, $4, $5)
+            ON CONFLICT (team_id, old_person_id) 
+            DO 
+                UPDATE SET override_person_id = $3, version = version + 1
+            RETURNING version, oldest_event
+            `,
+            [oldPerson.team_id, oldPerson.id, overridePerson.id, overridePerson.created_at, 1],
+            'personOverrides',
+            client
+        )
+
+        // (2)
+        // TODO: Race conditions here - 2 in parallel some might be left as the now overridden person?
```

Comment:
Enumerating some examples:

Say we have p1, p2, p3, p4

The we have these events
merge p1 into p2
merge p4 into p2
merge p2 into p3

We want to end up with:
p1 -> p3
p2 -> p3
p4 -> p3

If we process in order, this is fine.

If we process as:
merge p1 into p2
merge p2 into p3
merge p4 into p2

We'll get:
p1 -> p3
p2 -> p3

As it stands we'd fail to add p4 -> p3 as p2 no longer exists. But it's possible that we resolve that issue and add p4 -> p3 by looking up if p2 has already been merged into something else. We'd need to do that recursively as well I think 😬  Do we need to do this recursively anyway?

As for the concurrent case, we could go ahead and use Serializable isolation level. Or something else 🤔  will have a think. Probably missing something silly.